hibernate-core/src/main/java/org/hibernate/query/criteria/internal/predicate/InPredicate.java

```diff
@@ -197,14 +201,51 @@ public String render(boolean isNegated, RenderingContext renderingContext) {
 				}
 			}
 			else {
+				Expression lastValue = null;
 				buffer.append( '(' );
 				String sep = "";
 				for ( Expression value : values) {
+					lastValue = value;
 					buffer.append( sep )
 							.append( ( (Renderable) value )
 									.render( renderingContext ) );
 					sep = ", ";
 				}
+
+
+				if (renderingContext.getCriteriaLiteralHandlingMode() == LiteralHandlingMode.BIND &&
+						lastValue instanceof LiteralExpression) {
+
+					int bindValueCount = values.size();
+					int bindValueMaxCount = bindValueCount;
+
+					final SessionFactoryImplementor sfi = criteriaBuilder().getEntityManagerFactory().unwrap( SessionFactoryImplementor.class );
+					final Dialect dialect = sfi.getServiceRegistry().getService(JdbcServices.class).getJdbcEnvironment().getDialect();
+					final int inExprLimit = dialect.getInExpressionCountLimit();
+
+					boolean inClauseParameterPaddingEnabled = sfi.getSessionFactoryOptions().inClauseParameterPaddingEnabled() &&
+									bindValueCount > 2;
+
+					if ( inClauseParameterPaddingEnabled ) {
+						int bindValuePaddingCount = MathHelper.ceilingPowerOfTwo( bindValueCount );
+
+						if (inExprLimit > 0 && bindValuePaddingCount > inExprLimit ) {
+							bindValuePaddingCount = Math.max(bindValueCount, inExprLimit);
+						}
```

Comment:
I think we'd better leave the above logic in the other PR of yours (i.e. https://github.com/hibernate/hibernate-orm/pull/3468) to make concerns separated.

Reply:
So what do you suggest as a solution? 
I thought as Hibernate 6.0 is already in alpha state, and it is changing the generation of the sql completly, it would be a good enough solution to dublicate some code. 
Alternativ the InPredicate should put the values as list into one parameter so the QueryParameterBindingsImpl can do its job. But I do not know how to do this exactly.
Maybe it is enough to include my test case. So we can be sure this bug is fixed in release 6.0.

Reply:
I've thought this PR is targeting v5, not v6. I don't know too much about the implementation detail, but given you have created another PR for the above, including it here seems distracting.

Reply:
Two different issues, hence two different solutions.
#3468 is fixing [HHH-14109](https://hibernate.atlassian.net/projects/HHH/issues/HHH-14109). This works if the value list is set as parameter to the prepared statement. For Oracle that means padding values between 513 and 999 to 1000. 

This PR fixes using criteria query and using literalHandlingMode == BIND does not do any parameter padding. As shown by my extension to the testcase.
Yes this PR is targeting v5. But I thought code duplication is no big deal for v5. As far I understand v6 correct it will rewrite the hole sql generation code anyway. 

As I said in my comment to the PR. That is a somewhat simple solution. May be you just ake my testcase from the testcase to make sure the bug will be fixed in v6.

Reply:
I am not concerned with v6 since this PR is targeting v5. My concern is we should make ticket and PR independent. If the duplicated content is included, there are the following issues I can think of from ticket management:

1. we might end up with duplicated code review, which is a waste of time and efforts
2. if the duplicated content needs improvement, we need to modify multiple PRs, which is error-prone
3. if the duplicated content is not mandatory for this ticket (it seems so from my understanding), including it mislead the reviewer to miss the main point in the ticket.

Maybe I misunderstood something, but concern separating is a good thing in ticket management, just like in our routine coding practice.